docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+A new thinly-provisioned volume, based on the pool, will be created for running container’s.
+
+The kernel tracks the usage of the pool device at the block device layer. The usage here includes image layers and container’s writable layers.
+
+Since the kubelet has to track the writable layer usage anyways, we can subtract the aggregated root filesystem usage from the overall pool device usage to get the image layer’s disk usage.
+
+Linux quota and `du` will not work with device mapper.
+
+A docker dry run option (mentioned above) is another possibility.
+
+
+#### Container Writable Layer
+
+###### Overlayfs / Aufs
+
+Docker creates a separate directory for the container’s writable layer which is then overlayed on top of read-only image layers.
+
+Both the previously mentioned options of `du` and `Linux Quota` will work for this case as well.
+
+Kubelet can use `du` to track usage and enforce `limits` once disk becomes a schedulable resource. As mentioned earlier `du` is resource intensive.
+
+To use Disk quota, kubelet will have to allocate a separate gid per container. Kubelet can reuse the same gid for multiple instances of the same container (restart scenario). As and when kubelet garbage collects dead containers, the usage of the container will drop.
+
+If local disk becomes a schedulable resource, `linux quota` can be used to impose `request` and `limits` on the container writable layer.
+`limits` can be enforced using hard limits. Enforcing `request` will be tricky. One option is to enforce `requests` only when the disk availability drops below a threshold (10%). Kubelet can at this point evict pods that are exceeding their requested space. Other options include using `soft limits` with grace periods, but this option is complex.
+
+###### Devicemapper
+
+Accessing stats requires entering the docker’s mount namespace and then fetch the filesystem stats on the fs mounted for the container’s writable layer.
+
+For example, if the container’s id is `92b94ac8c3`, then to get disk usage stats for that container, the following command needs to be run,
+
+```shell
+ nsenter -m -t 16663 stat -f /var/lib/docker/devicemapper/mnt/92b94ac8c3f3a932f6fdf30b86edcdff94d8e5f71e36b79da8419294c19eb6f2/rootfs/
+
+File: "/var/lib/docker/devicemapper/mnt/92b94ac8c3f3a932f6fdf30b86edcdff94d8e5f71e36b79da8419294c19eb6f2/rootfs/"
+
+    ID: e1cef0d95ec0beb7 Namelen: 255     Type: ext2/ext3
+
+Block size: 4096       Fundamental block size: 4096
+
+Blocks: Total: 25770312   Free: 25754754   Available: 24439938
+
+Inodes: Total: 6553600    Free: 6553556
+
+```
+
+*Note: The container’s writable layer is mounted only when the container is running.*
+
+To enforce `limits` the volume created for the container’s writable layer filesystem can be dynamically [resized](https://jpetazzo.github.io/2014/01/29/docker-device-mapper-resize/), to not use more than `limit`. `request` will have to be enforced by the kubelet.
+
```

Comment:
As container rootfs is a block device, container size is limited by size of block device. Right now docker default is 100G per container but this is tunable with a docker option dm.basesize=X during first time start of docker. So that should allow you to come up with a common size limit per container.

We have been thinking of implementing something where one could specify bigger size for a freshly created container. So say default is 10G then one could say that create a container with size 50G. But this is all theory right now. Once patches are implemented, we will know if upstream likes these or not.

The article you have linked is all the magic behind dokcer's back. I would rather propose proper API/functionality in docker so that one could do what's needed instead of going behind docker's back and start playing with its metadata files or thin pool  or thin devices etc.